[Jerome Hoo] iP

jeromehjj opened the pull request:
# Violet
>“People often say that motivation doesn’t last long. Well, neither does bathing – that’s why we recommend it daily.” – Zig Ziglar ([source](https://dansilvestre.com/productivity-quotes/#5_People_often_say_that_motivation_doesnt_last_long_Well_neither_does_bathing_-_thats_why_we_recommend_it_daily_-_Zig_Ziglar))

Violet frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,
1. download it from [here](https://github.com/jeromehjj/ip/releases/tag/A-Jar)
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉 

And it is **FREE**!

Features:
- [x] Managing tasks
- [x] Managing deadlines
- [x] Managing events 
- [x] Managing todo/s
##
If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public static void main(String[] args) {
        Application.launch(Main.class, args);
    }
```

Commit:
Add Gradle support

Commit:
Added Level-1

Commit:
Added Level-2

Commit:
Added Level-3

Commit:
Added Level-4

Commit:
Added A-TextUiTesting

Commit:
Added Level-5

Commit:
Added Level-6

Commit:
Duke class: fix and handle exceptions

Commit:
EXPECTED.TXT: Update expected results

Commit:
Clean up code from previous week

Commit:
Add Level-7: Save functionality

Commit:
Ignore files in data/

Commit:
Add Level-8 to store date and time in deadline and event objects

Commit:
Ignore files in data/

Commit:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Duke.java

Resolution:
Keep one since conflicts are identical

Commit:
Fix bugs for saving files

Commit:
Merge remote-tracking branch 'upstream/add-gradle-support'

Commit:
Add more OOP and package all classes into one package

Commit:
Add JUnit tests

Commit:
Add find function

Commit:
Modify code to follow Coding Standard

Commit:
Add JavaDoc comments to the code

Commit:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Deadline.java
#	src/main/java/duke/Parser.java
#	src/main/java/duke/Storage.java
#	src/main/java/duke/Task.java
#	src/main/java/duke/TaskList.java
#	src/main/java/duke/ToDo.java
#	src/main/java/duke/Ui.java

Resolution :
Manually fix conflicts

Commit:
Merge tag 'Level-9'

Commit:
Add find functionality and JavaDoc

ChongCheeKaiClarence commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,255 @@
+    public boolean readCommand(TaskList taskList) {
+        Scanner sc = new Scanner(command);
+        String input = sc.nextLine();
+        String[] inputArray = input.split(" ");
```
Maybe u could name the array as inputs?

ChongCheeKaiClarence commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,145 @@
+package duke;
+import java.io.File;
```
I think you could use line breaks to show the distinction between packages.

ChongCheeKaiClarence commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,145 @@
+     * @throws FileNotFoundException If file does not exist.
+     */
+    public static void formatToList(File f, ArrayList<Task> taskList) throws FileNotFoundException {
+        Scanner s = new Scanner(f);
```
Maybe you should use scanner instead of s, and file instead f. Easier to keep track of variables.

ChongCheeKaiClarence commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,66 @@
+     * @return the task of given index found in list.
+     */
+    public Task get(int index) {
+        Task t = tasks.get(index);
```
Maybe you could return directly?

ChongCheeKaiClarence commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,63 @@
+    }
+
+    public String getStatusIcon() {
+        return (this.isDone ? "X" : " ");
```
Are the brackets necessary?

ChongCheeKaiClarence reviewed (commented):
Looks good.
Maybe u could try packaging some of the classes up.

ChongCheeKaiClarence commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,255 @@
+        String[] inputArray = input.split(" ");
+        String first = inputArray[0];
+
+        try {
```
You can consider using a switch statement.

comicalromance commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,51 @@
+    /**
+     * Prints the greeting message when Duke is opened.
+     */
+    public static void greeting() {
```
Could this function be renamed / rephrased as a verb (e.g. printGreeting?)

comicalromance commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,145 @@
+                    date += taskDescription[3];
+                }
+
+                if (taskType.equals("T")) {
+                    task = new ToDo(description, isDone);
+                }
+
+                if (taskType.equals("D")) {
+                    task = new Deadline(description, isDone, date);
+                }
+
+                if (taskType.equals("E")) {
+                    task = new Event(description, isDone, date);
+                }
```
Could this be rewritten as a switch statement instead?

comicalromance commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,145 @@
+
+            printFileContents(this.filePath);
+            formatToList(file, taskList);
+        } catch (FileNotFoundException e) {
+            throw new DukeException("Error: Cannot load file!");
+        } catch (IOException e) {
+            throw new DukeException("Error: File does not exist");
+        }
```
I really like how you handled multiple types of exceptions and returned them as corresponding DukeExceptions for better bookkeeping!

comicalromance commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,255 @@
+     *
+     * @param taskList A collection of Task objects.
+     */
+    public static void listTasks(TaskList taskList) {
```
Just curious if it is possible for this method (as well as the other Parser methods) to be private and non-static, as it only seems to be referenced in the readCommand method which is in Parser anyway.

comicalromance commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,74 @@
+        try {
+            this.event = LocalDateTime.parse(at.substring(1));
+        } catch (DateTimeParseException e) {
+            throw new DukeException("Please input the deadline in yyyy-mm-ddTHours:Minutes:Seconds format. "
```
I like how the specific input for the time format is specified for a more user-friendly experience! (but I think 'deadline' should be 'event time' or something similar instead)

comicalromance reviewed (commented):
LGTM! There were a few user interactions (e.g. println and readLine) in your Parser class that I believe could be given to the UI class to handle, and some bits of code in the constructors are duplicated and can be simplified a bit (e.g. using the this() keyword to call its constructor). The code style however was really great and there were close to no issues with it!

rexong commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    @Override
+    public String fileStatus() {
```
`fileStatus` sounds like a noun to me. Perhaps you can consider `showFileStatus` or `printFileStatus`?

rexong commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,255 @@
+     * @param taskList A collection of Task objects.
+     * @throws DukeException If input given is invalid.
+     */
+    public static void unmarkTask(String input, TaskList taskList) throws DukeException {
```
I really like how you name your methods in this class, very specific as to what it actually does. 👍

rexong commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,145 @@
+     * @return The collection of Tasks that was saved in the file.
+     * @throws DukeException If file does not exist.
+     */
+    public ArrayList<Task> load() throws DukeException {
```
Probably can consider naming this as `loadFileToDuke` to make the method name more meaningful.

rexong commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,145 @@
+     *
+     * @param taskList The collection of Task objects.
+     */
+    public void save(TaskList taskList) {
```
Same as the `load` method name above, probably naming it `saveToFile` is more intuitive.

rexong commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,51 @@
+public class Ui {
+    private static final String LINE = "____________________________________________________________\n";
+    /** Scanner to read input from the user */
+    private Scanner sc;
```
Maybe can name it as `scanner` instead of `sc` to make it more intuitive.

rexong commented on src/test/java/duke/DeadlineTest.java:
```diff
@@ -0,0 +1,21 @@
+package duke;
+
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.fail;
```
I think importing static should go before the regular import statement. 
```suggestion
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.fail;

import org.junit.jupiter.api.Test;
```

rexong reviewed (commented):
I think your code looks good. It follows most of the coding quality stated in the module. Just have some suggestions regarding how you name your methods and attributes to make it more intuitive. 😄

Commit:
Add checkstyle

Commit:
Add GUI to Duke

Commit:
Add assertions to execute() methods

Add assertions for runtime to verify them in methods in certain
classes.

jeromehjj committed:
Merge pull request #2 from jeromehjj/branch-A-Assertions

Add assertions to execute() methods

Commit:
Improve code quality of methods in all classes

The code quality in most classes follow the coding quality
standard.

Improving the code quality across all methods in every class
will help to increase the readability and quality of code.

Commit:
Merge branch 'branch-A-CodeQuality'

Commit:
Add Priority to tasks

Commit:
Improve GUI

Commit:
Edit Main.java to include title

Commit:
Add User Guide and Ui.png
